gh-413 - refactor createFromDataElements

abwilson23 opened the pull request:
Just make sure creating a data spec from data elements still functions correctly in the ui. 

closes #413

abwilson23 committed:
gh-413 - refactor createFromDataElements

NigelPalmer commented on src/app/data-explorer/data-specification.service.ts:
```diff
@@ -542,34 +541,44 @@ export class DataSpecificationService {
-        if (!rootDataModel || !rootDataModel.id) {
-          return throwError(() => new Error('No root data model'));
+        if (!rootDataModel?.id || !dataSpecification?.id) {
+          return throwError(() => new Error('Missing root data model or data specification'));
```
I don't like this. This makes it harder to work out what has gone wrong. Is it broken because the root model is missing or because the data specification is missing? I would prefer this check as separate errors: e.g:
```
        if (!rootDataModel?.id) {
          return throwError(() => new Error('Missing root data model'));

        if (!dataSpecification?.id) {
          return throwError(() => new Error('Missing data specification'));
```

abwilson23 replied:
Reverted.

NigelPalmer commented on src/app/data-explorer/data-specification.service.spec.ts:
```diff
@@ -484,12 +484,7 @@ describe('DataSpecificationService', () => {
-      submissionSDEServiceStub.mapToDataSpecification.mockImplementation((dataModel) => {
-        expect(dataModel.id).toBe(dataSpecification.id);
-        return dataSpecification;
-      });
```
Not clear as to why this has been removed from the test?

abwilson23 replied:
Neither..! Added back.

NigelPalmer commented on src/app/data-explorer/data-specification.service.ts:
```diff
@@ -542,34 +541,44 @@ export class DataSpecificationService {
+
+        // We don't care about the output above so just return the data specification.
+        return this.submissionSDEService.mapToDataSpecification(populatedDataSpecification);
+      })
```
This is not working? When I try and create a specification and go to look at the query builder I get the error:
```
"Mauro Data Explorer - Query Builder Core Table" (data model profile) for DataModel validation errors:
- This field cannot be empty
- Format needs to be [Schema].[Table] e.g: people.patient
```
I checked the network log and I can see the profile being "got", but I cannot see it being saved.

abwilson23 replied:
This is due to the fact that I was incorrectly using `map` here. The fix is to use `switchMap` to switch to the observable emitted by the `saveQueryBuilderCoreTableProfile()` method and then pipe off the end of that to map the dataSpecification properly. This ensures the save call completes successfully before moving along to the next step.

NigelPalmer reviewed (changes requested):
A few coding issues and I couldn't get it to work?

abwilson23 committed:
gh-412 - address pr feedback.

- Use switchMap and pipe into result to ensure completion of saveQueryBuilderProfile call.
- Make errors clearer.
- Fix tests.

NigelPalmer committed:
Merge branch 'develop' into feature/gh-413

NigelPalmer reviewed (approved):
This now all seems to be working as it should.
